content/blog/aws-serverless-analytics/index.md

```diff
@@ -0,0 +1,604 @@
+
+1. Scale the read and write workloads independently. Redshift was overprovisioned to handle the mismatch of a continuous write workload at peak hours. A decrease in operating cost was the result of adopting this architecture.
+2. Pay as you go. Firehose, S3, Lambda, and Athena are all billed based on usage. Kinesis is the exception charging per shard hour, which for most deployments would be provisioned for peak load.
+3. Configurable latency. Firehose can dump batches of data in intervals of 1-15 minutes. If low latency is a concern, you can achieve it (see the end for caveats on compaction). We don’t have to worry about an increased write workload impacting reads.
+4. Malformed records won’t block the data ingestion pipeline. Redshift COPY has a [maxerror](https://docs.aws.amazon.com/redshift/latest/dg/copy-parameters-data-load.html#copy-maxerror) that allows skipping over bad records, but there isn’t any automation to isolate and alert on those records. Firehose writes unsuccessful events to a separate directory and allows processing to continue.
+
+While our new serverless streaming architecture brings a slew of benefits that enable ease of operation and the next level of scale, there are some drawbacks. For one, there is substantially more infrastructure to configure than the single Redshift cluster. How do we implement this in a repeatable fashion?
+
+## Architecture as Code with Pulumi
+
+We want repeatable infrastructure that we can reliably develop, test, and deploy. Creating a single Redshift instance through the console was manageable, but our serverless architecture utilizes multiple instances of five AWS services. Serverless solutions frequently require many more granular resources, and thus the complexity of the infrastructure as code increases, and we need more robust tools to manage this complexity - just like we manage software complexity.
+
+Pulumi takes care of the complexity by bringing software development best practices like abstraction, encapsulation, and modularity to infrastructure. You can use design patterns in your favorite programming language to provision modern cloud infrastructure. [Component Resources](https://www.pulumi.com/blog/creating-and-reusing-cloud-components-using-package-managers/) are our tools for encapsulating low-level cloud infrastructure into Architecture as Code that is shareable across teams.
```

Comment:
```suggestion
Pulumi takes care of the complexity by bringing software development best practices like abstraction, encapsulation, and modularity to infrastructure. You can use design patterns in your favorite programming language to provision modern cloud infrastructure. [Component Resources]({{< rel "/blog/creating-and-reusing-cloud-components-using-package-managers/index.md" >}}) are our tools for encapsulating low-level cloud infrastructure into Architecture as Code that is shareable across teams.
```
Use Hugo shortcode for links to internal

Reply:
👍